Update Readme.md (#9)
riscv/riscv-edk2-platforms: Platform/SiFive/U5SeriesPkg/Readme.md

# Introduction of SiFive U5 Series Platforms
U5SeriesPkg provides the common EDK2 libraries and drivers for SiFive U5 series platforms. Currently the supported
platforms are Freedom U500 VC707 platform and Freedom U540 HiFive Unleashed platform.

Both platforms are built with below common edk2 packages under edk2-platforms repository,
- [**U5SeriesPkg**](https://github.com/tianocore/edk2-platforms/tree/master/Platform/SiFive/U5SeriesPkg)
- [**RiscVPlatformPkg**](https://github.com/tianocore/edk2-platforms/tree/master/Platform/RISC-V/PlatformPkg)
- [**RiscVProcessorPkg**](https://github.com/tianocore/edk2-platforms/tree/master/Silicon/RISC-V/ProcessorPkg)

## U500 Platform
This is a sample RISC-V EDK2 platform package used against to SiFive Freedom U500
VC707 FPGA Dev Kit, please refer to "SiFive Freedom U500 VC707 FPGA Getting
Started Guide" on https://www.sifive.com/documentation.
The binary built from Platform/SiFive/U5SeriesPkg/FreedomU500VC707Board can run on U500 VC707 FPGA board.
```
build -a RISCV64 -t GCC5 -p Platform/SiFive/U5SeriesPkg/FreedomU500VC707Board/U500.dsc
```

## U540 Platform

> This is a sample RISC-V EDK2 platform package used against to SiFive Freedom U540 HiFive Unleashed development board, please refer to "SiFive Freedom U540-C000 Manual" on https://www.sifive.com.

Reviewer: ```suggestion
This is a sample RISC-V EDK2 platform package used for the SiFive Freedom U540 HiFive Unleashed development board, please refer to "SiFive Freedom U540-C000 Manual" on https://www.sifive.com.

```

Author: done

The binary built from Platform/SiFive/U5SeriesPkg/FreedomU540HiFiveUnleashedBoard/ can run on both hardware and

> the main stream [QEMU](https://git.qemu.org/?p=qemu.git;a=summary).

Reviewer: ```suggestion
[QEMU](https://git.qemu.org/?p=qemu.git;a=summary). We confirmed that version 5.0 of QEMU can boot the firmware.

```

Author: done

Author: > . We confirmed that version 5.0 of QEMU can boot the firmware.

It is confirmed that version 5.0 of QEMU can boot the firmware, instead of "we".

Reviewer: Okay. I also thought about this haha

```
build -a RISCV64 -t GCC5 -p Platform/SiFive/U5SeriesPkg/FreedomU540HiFiveUnleashedBoard/U540.dsc
```
For running U540 edk2 binary on QEMU, use qemu-system-riscv64 under riscv64-softmmu to launch the binary with
below parameters,

```

> -cpu sifive-u54 -machine sifive_u -bios U540.fd -m 4096 -serial /your/console -nographic -smp cpus=5,maxcpus=5

Reviewer: ```suggestion
qemu-system-riscv64 -cpu sifive-u54 -machine sifive_u -bios U540.fd -m 2048 -nographic -smp cpus=5,maxcpus=5

```

Then people can quickly copy the command without thinking ;)
I have always used 2048M memory and it was fine.
Serial is not necessary and then people need to know how to connect to that.

Author: done

FPGA to EFI shell with console in/out enabled.

**FreedomU540HiFiveUnleashedBoard**
Currently the binary built from U540 edk2 package can boot SiFive Freedom U540 HiFive Unleashed
to EFI shell with console in/out enabled and Linux kernel. Please refer to https://github.com/riscv/riscv-uefi-edk2-docs for booting to Linux kernel.

## Supported Operating Systems
Please refer to https://github.com/riscv/riscv-uefi-edk2-docs.

## Known Issues and Limitations
Only RISC-V RV64 is verified on this platform.

## Related Materials
- [RISC-V UEFI Documents](https://github.com/riscv/riscv-uefi-edk2-docs)
- [RISC-V OpenSbi](https://github.com/riscv/opensbi)
- [SiFive U500 VC707 FPGA Getting Started Guide](https://sifive.cdn.prismic.io/sifive%2Fc248fabc-5e44-4412-b1c3-6bb6aac73a2c_sifive-u500-vc707-gettingstarted-v0.2.pdf)
- [SiFive Freedom U540-C000 Manual](https://sifive.cdn.prismic.io/sifive%2F834354f0-08e6-423c-bf1f-0cb58ef14061_fu540-c000-v1.0.pdf)
- [SiFive RISC-V Core Document](https://www.sifive.com/documentation)

## U5SeriesPkg Libraries and Drivers
### PeiCoreInfoHobLib
This is the library to create RISC-V core characteristics for building up RISC-V related SMBIOS records to support

> the single boot loader and OS image.

Reviewer: What do you mean? Why single?

Author: I found some people they like to use "single" image instead of "unified" image.
how about this,
This is the library to create RISC-V core characteristics for building up RISC-V related SMBIOS records to support
the single boot loader and OS image on the RISC-V variants.

Reviewer: ```suggestion
a single boot loader and OS image for all RISC-V platforms by discovering system configuration dynamically.
```

Author: that is not for system configuration, how about this

a single boot loader and OS image for all RISC-V platforms by discovering RISC-V hart configurations dynamically.

Reviewer: right, hart config not system config. I was trying to find the right word.
That's okay. If you pick that `Reviewed-By: Daniel Schaefer <daniel.schaefer@hpe.com>` ;)

This library leverage the silicon libraries provided in Silicon/SiFive.

### RiscVPlatformTimerLib
This is common U5 series platform timer library which has the platform-specific timer implementation.

### SerialLib
This is common U5 series platform serial port library.

### TimerDxe
This is common U5 series platform timer DXE driver which has the platform-specific timer implementation.

## U500 Platform Libraries and Drivers
### RiscVOpensbiPlatformLib
In order to reduce the dependencies with RISC-V OpenSBI project (https://github.com/riscv/opensbi) and fewer
burdens to EDK2 build process, the implementation of RISC-V EDK2 platform is leveraging platform source code from
OpenSBI code tree. The "platform.c" under OpenSbiPlatformLib is cloned from

> [RISC-V OpenSBI code tree](Silicon/RISC-V/ProcessorPkg/Library/RiscVOpensbiLib/opensbig) and built based on edk2

Reviewer: ```suggestion
[RISC-V OpenSBI code tree](Silicon/RISC-V/ProcessorPkg/Library/RiscVOpensbiLib/opensbi) and built based on edk2

```

Author: haha opensbig

initialization.

## U540 Platform Libraries and Drivers
### RiscVOpensbiPlatformLib
In order to reduce the dependencies with RISC-V OpenSBI project (https://github.com/riscv/opensbi) and fewer
burdens to EDK2 build process, the implementation of RISC-V EDK2 platform is leveraging platform source code from
OpenSBI code tree. The "platform.c" under OpenSbiPlatformLib is cloned from
[RISC-V OpenSBI code tree](Silicon/RISC-V/ProcessorPkg/Library/RiscVOpensbiLib/opensbig) and built based on edk2
build environment.

### PlatformPei
This is the platform-implementation specific library which is executed in early PEI phase for U540 platform
initialization.

## U5SeriesPkg Platform PCD settings

| **PCD name** |**Usage**|
|----------------|----------|
|PcdU5PlatformSystemClock| U5 series platform system clock|
|PcdNumberofU5Cores| Number of U5 core enabled on U5 series platform|
|PcdE5MCSupported| Indicates whether the Monitor Core (E5) is supported on U5 series platform|
|PcdU5UartBase|Platform serial port base address|

## Platform Owners
Chang, Abner <abner.chang@hpe.com>
Chen, Gilbert <gilbert.chen@hpe.com>

> Daniel Schaefer <daniel.schaefer@hpe.com>

Reviewer: Either reverse your names or mine.

Author: done
